### PR TITLE
TST: add test for empty frame groupby dtypes consistency

### LINKED ISSUE #20888: Groupby on empty frame different results based on dtypes.
#### Code Sample, a copy-pastable example if possible

```python
In [117]: import pandas as pd

In [118]: df = pd.DataFrame({"a":[1], "b":[2], "c":[3], "d":[4]})

In [119]: group_keys = ["a", "b", "c"]

In [120]: g = df[df.a==2].groupby(group_keys)

In [121]: g.first().index
Out[121]:
MultiIndex(levels=[[], [], []],
           labels=[[], [], []],
           names=[u'a', u'b', u'c'])
# Change the dtype for "d"
In [122]: df = pd.DataFrame({"a":[1], "b":[2], "c":[3], "d":["d"]})

In [123]: g = df[df.a==2].groupby(group_keys)

In [124]: g.first().index
Out[124]: Index([], dtype='object')

# Version 0.18
In [36]: g.first().index
Out[36]:
MultiIndex(levels=[[], [], []],
           labels=[[], [], []],
           names=[u'a', u'b', u'c'])

```
#### Problem description

The groupby should return multi-index in both cases.  In one case when we have all int64 data, the resultant groupby on an empty frame returns the multi-index.  When we have a str (object) dtype, the return is an empty index.  Not consistent.  Previous version, 0.18.0 returned multi-index in both cases.

#### Expected Output
```python
In [36]: g.first().index
Out[36]:
MultiIndex(levels=[[], [], []],
           labels=[[], [], []],
           names=[u'a', u'b', u'c'])
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.14.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.36.3.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: C
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: 3.5.0
pip: 9.0.3
setuptools: 39.0.1
Cython: 0.28.2
numpy: 1.14.2
scipy: 1.0.1
pyarrow: 0.9.0
xarray: 0.10.2
IPython: 5.6.0
sphinx: 1.7.2
patsy: 0.5.0
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: 2.5.2
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.2.1
bs4: 4.3.2
html5lib: 0.999
sqlalchemy: 1.2.6
pymysql: None
psycopg2: 2.7.4 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by ganevgv**
- [x] closes #20888
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/test_multilevel.py [hunk 1]
```diff
@@ -1490,6 +1490,19 @@ def test_frame_dict_constructor_empty_series(self):
...
+        group_keys = ["a", "b", "c"]
+        df = DataFrame({"a": [1], "b": [2], "c": [3], "d": [4]})
+        g = df[df.a == 2].groupby(group_keys)
+        expected = g.first().index
+
+        df = DataFrame({"a": [1], "b": [2], "c": [3], "d": ["d"]})
+        g = df[df.a == 2].groupby(group_keys)
+        result = g.first().index
+
+        tm.assert_index_equal(result, expected)
```
Comment (gfyoung) [inline pandas/tests/test_multilevel.py]: Actually, let's do a parameterization over `d` as follows:

~~~python
def test_empty_frame_groupby_dtypes_consistency(self, d):
   group_keys = [...]
   df = DataFrame({"a": [1], "b": [2], "c": [3], "d": [d]})
   g = df[df.a == 2].groupby(group_keys)

   expected = ...  # Construct the expected result
   result = g.first().index

   tm.assert_index_equal(result, expected)
~~~

File: pandas/tests/test_multilevel.py [hunk 2]
```diff
@@ -1490,6 +1490,16 @@ def test_frame_dict_constructor_empty_series(self):
         DataFrame({"foo": s1, "bar": s2, "baz": s3})
         DataFrame.from_dict({"foo": s1, "baz": s3, "bar": s2})
 
+    @pytest.mark.parametrize("d", [4, "d"])
+    def test_empty_frame_groupby_dtypes_consistency(self, d):
+        # GH 20888
+        group_keys = [...]
```
Comment (gfyoung) [inline pandas/tests/test_multilevel.py]:
```suggestion
        group_keys = ["a", "b", "c"]
```

I used ellipses as shorthand to indicate you use `group_keys = ["a", "b", "c"]`

File: pandas/tests/test_multilevel.py [hunk 3]
```diff
@@ -1490,6 +1490,16 @@ def test_frame_dict_constructor_empty_series(self):
...
+        df = DataFrame({"a": [1], "b": [2], "c": [3], "d": [d]})
+        g = df[df.a == 2].groupby(group_keys)
```
Comment (gfyoung) [inline pandas/tests/test_multilevel.py]: Nit: add new line above this

File: pandas/tests/test_multilevel.py [hunk 4]
```diff
@@ -1490,6 +1490,16 @@ def test_frame_dict_constructor_empty_series(self):
...
+        result = g.first().index
+        expected = Index([], dtype=int)
+        tm.assert_index_equal(result, expected)
```
Comment (gfyoung) [inline pandas/tests/test_multilevel.py]: Nit: add new line above this one

File: pandas/tests/test_multilevel.py (hunk 1 above)
Comment (ganevgv) [reply pandas/tests/test_multilevel.py]: Thanks for the feedback, @gfyoung! Changes done.

File: pandas/tests/test_multilevel.py (hunk 2 above)
Comment (ganevgv) [reply pandas/tests/test_multilevel.py]: Sorry - my bad. I put back the correct `group_keys`.

File: pandas/tests/test_multilevel.py (hunk 3 above)
Comment (ganevgv) [reply pandas/tests/test_multilevel.py]: done

File: pandas/tests/test_multilevel.py (hunk 4 above)
Comment (ganevgv) [reply pandas/tests/test_multilevel.py]: done

**Comment by gfyoung** [comment]
Thanks @ganevgv !

**Pull request merged**